spec/message.abnf

```diff
@@ -1,18 +1,30 @@
-message = pattern / complex-message
+message = simple-message / complex-message
 
-complex-message = "{{" [s] *(declaration [s]) body [s] "}}"
+simple-message = [almost-keyword]
+               / ((almost-keyword not-keyword / text-start / expression) pattern)
+complex-message = *(declaration [s]) body
```

Comment:
Should we still allow for non-space-separated declarations? As in, should this still be valid?
```
.input{$foo}.input{$bar}.match{$foo}one{{o-n-e}}*{{e-l-s-e}}
```

For the sake of legibility, I'd really like to require at least a space separating each declaration, `.match` and variant:
```
.input{$foo} .input{$bar} .match{$foo} one{{o-n-e}} *{{e-l-s-e}}
```

Reply:
I don't think those spaces should be required if the grammar is unambiguous without them. Legibility is for people, not machines.

Reply:
@gibson042 wrote:
> I don't think those spaces should be required if the grammar is unambiguous without them. Legibility is for people, not machines.

@eemeli wrote:
> For the sake of legibility, I'd really like to require at least a space separating each declaration, .match and variant.

I'm strongly on @eemeli's side. Legibility is for people. But requiring the spaces will make sure that even machine-generated messages are legible to people. And it will make sure that there's a consistent "custom" of writing messages with spaces. (You'd be surprised how Japanese, in particular unexperienced students, tend to omit the spaces because they are not used to spaces in Japanese text.)

Reply:
I disagree that messages like `.input{$foo} .input{$bar} .match{$foo} one{{o-n-e}} … *{{e-l-s-e}}` are sufficiently legible to people, and in my own use would use additional separating line feeds and spaces:
```
.input {$foo}
.input {$bar}
.match {$foo}
one {{o-n-e}}
…
* {{e-l-s-e}}
```
Insisting on unnecessary spaces at a syntactic level is a half-measure that doesn't actually address any legibility concern IMO.

Reply:
@gibson042 I'd definitely also use line feeds in some places. I'd be fine if we require line feeds in these positions. Let's do this as a full measure rather than a half-measure.

Reply:
We expect many messages to only exist in a single line form, e.g. in spreadsheets. We should not require line breaks. As for spaces -- we can recommend a style, but I strongly believe we should be lenient in what we parse, and be as whitespace-agnostic as possible.

Reply:
@duerst 

We long ago decided about non-essential whitespace. Resource formats and message authors can impose additional formatting and structure onto messages, but the syntax itself does not impose it. Messages can often appear in a single, simple string (think Java properties file), where line breaks would be a problem or be reduced to illegible escape sequences (reducing rather than adding value).

Reply:
During today's call we resolved to not add any new whitespace requirements.